File: api/typescript/src/core/lib/services/service_identifiers.ts
import {ServiceName, ServiceUUID} from "./service";
import kurtosis_core_rpc_api_bindings =  require("../../kurtosis_core_rpc_api_bindings/api_container_service_pb")

const VALID_UUID_MATCHES_ALLOWED = 1

// Docs available at https://docs.kurtosis.com/sdk#service-identifiers
export class ServiceIdentifiers {
    public readonly serviceNameToUuids: Map<ServiceName, ServiceUUID[]>;
    public readonly serviceUuids: Map<ServiceUUID, boolean>;
    public readonly serviceShortenedUuidToUuids: Map<string, ServiceUUID[]>;

    constructor(historicalIdentifiers : kurtosis_core_rpc_api_bindings.ServiceIdentifiers[]) {
        this.serviceUuids = new Map<string, boolean>();
        this.serviceNameToUuids = new Map<ServiceName, ServiceUUID[]>();
        this.serviceShortenedUuidToUuids = new Map<string, ServiceUUID[]>();
        historicalIdentifiers.forEach((serviceIdentifiers) => {
            let serviceName = serviceIdentifiers.getName();
            let serviceUuid : ServiceUUID = serviceIdentifiers.getServiceUuid();
            let shortenedUuid = serviceIdentifiers.getShortenedUuid();

            this.serviceUuids.set(serviceUuid, true);

            if (!(serviceName in this.serviceNameToUuids)) {
                this.serviceNameToUuids.set(serviceName, []);
            }
            let serviceUuids = this.serviceNameToUuids.get(serviceName)!
            serviceUuids.push(serviceUuid)
            this.serviceNameToUuids.set(serviceName, serviceUuids)

            if (!(shortenedUuid in this.serviceShortenedUuidToUuids)) {
                this.serviceShortenedUuidToUuids.set(shortenedUuid, []);
            }
            serviceUuids = this.serviceShortenedUuidToUuids.get(shortenedUuid)!
            serviceUuids.push(serviceUuid)
            this.serviceShortenedUuidToUuids.set(shortenedUuid, serviceUuids)
        });
    }

    public getServiceUuidForIdentifier(identifier: string): ServiceUUID {
        if (this.serviceUuids.has(identifier)) {
            return identifier as ServiceUUID
        }

        if (this.serviceShortenedUuidToUuids.has(identifier)) {
            let matches = this.serviceShortenedUuidToUuids.get(identifier)!
            if (matches.length === VALID_UUID_MATCHES_ALLOWED) {
                return matches[0];
            } else if (matches.length > VALID_UUID_MATCHES_ALLOWED) {
                throw new Error(`Found multiple services ${matches} matching shortened uuid ${identifier}. Please use a uuid to be more specific`)
            }
        }

        if (this.serviceNameToUuids.has(identifier)) {
            let matches = this.serviceNameToUuids.get(identifier)!
            if (matches.length === VALID_UUID_MATCHES_ALLOWED) {
                return matches[0];
            } else if (matches.length > VALID_UUID_MATCHES_ALLOWED) {
                throw new Error(`Found multiple services ${matches} matching name ${identifier}. Please use a uuid to be more specific`)
            }
        }

        throw new Error(`No matching uuid for identifier ${identifier}`)
    }

    public getOrderedListOfNamesAndUuids(): String[] {
        let serviceNames: string[] = [];
        let serviceUuids: string[] = [];

        for (let name in this.serviceNameToUuids) {
            serviceNames.push(name);
        }

        for (let uuid in this.serviceUuids) {
            serviceUuids.push(uuid);
        }

        serviceNames.sort();
        serviceUuids.sort();

        return [...serviceNames, ...serviceUuids];
    }
}

File: api/golang/core/lib/services/service_identifiers.go
package services

import (
	"github.com/kurtosis-tech/kurtosis/api/golang/v2/core/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/stacktrace"
	"sort"
)

const (
	validUuidMatchesAllowed = 1
)

// Docs available at https://docs.kurtosis.com/sdk#service-identifiers
type ServiceIdentifiers struct {
	serviceNameToUuids          map[ServiceName][]ServiceUUID
	serviceUuids                map[ServiceUUID]bool
	serviceShortenedUuidToUuids map[string][]ServiceUUID
	enclaveNameForLogging       string
}

func NewServiceIdentifiers(enclaveNameForLogging string, allIdentifiers []*kurtosis_core_rpc_api_bindings.ServiceIdentifiers) *ServiceIdentifiers {
	serviceNames := map[ServiceName][]ServiceUUID{}
	serviceUuids := map[ServiceUUID]bool{}
	serviceShortenedUuid := map[string][]ServiceUUID{}

	for _, serviceIdentifiers := range allIdentifiers {
		serviceName := ServiceName(serviceIdentifiers.GetName())
		serviceUuid := ServiceUUID(serviceIdentifiers.GetServiceUuid())
		serviceUuids[serviceUuid] = true
		shortenedUuid := serviceIdentifiers.GetShortenedUuid()
		serviceNames[serviceName] = append(serviceNames[serviceName], serviceUuid)
		serviceShortenedUuid[shortenedUuid] = append(serviceShortenedUuid[shortenedUuid], serviceUuid)
	}

	return &ServiceIdentifiers{enclaveNameForLogging: string(enclaveNameForLogging), serviceNameToUuids: serviceNames, serviceUuids: serviceUuids, serviceShortenedUuidToUuids: serviceShortenedUuid}
}

func (identifiers *ServiceIdentifiers) GetServiceUuidForIdentifier(identifier string) (ServiceUUID, error) {
	if _, found := identifiers.serviceUuids[ServiceUUID(identifier)]; found {
		return ServiceUUID(identifier), nil
	}

	if matches, found := identifiers.serviceShortenedUuidToUuids[identifier]; found {
		if len(matches) == validUuidMatchesAllowed {
			return matches[0], nil
		} else if len(matches) > validUuidMatchesAllowed {
			return "", stacktrace.NewError("Found multiple services '%v' matching shortened uuid '%v' in enclave '%v'. Please use a uuid to be more specific", matches, identifier, identifiers.enclaveNameForLogging)
		}
	}

	if matches, found := identifiers.serviceNameToUuids[ServiceName(identifier)]; found {
		if len(matches) == validUuidMatchesAllowed {
			return matches[0], nil
		} else if len(matches) > validUuidMatchesAllowed {
			return "", stacktrace.NewError("Found multiple services '%v' matching name '%v' in enclave '%v'. Please use a uuid to be more specific", matches, identifier, identifiers.enclaveNameForLogging)
		}
	}

	return "", stacktrace.NewError("No matching uuid for identifier '%s'", identifier)
}

func (identifiers *ServiceIdentifiers) GetOrderedListOfNames() []string {
	var serviceNames []string
	for name := range identifiers.serviceNameToUuids {
		serviceNames = append(serviceNames, string(name))
	}
	sort.Strings(serviceNames)
	return serviceNames
}
